syncr_backend/external_interface/dht_util.py

```diff
@@ -0,0 +1,122 @@
+from typing import List
+from typing import Tuple
+
+from kademlia.network import Server  # type: ignore
+from kademlia.storage import IStorage  # type: ignore
+
+from syncr_backend.constants import TRACKER_DROP_AVAILABILITY_TTL
+
+_node_instance = None
+
+
+def get_dht(
+    bootstrap_ip_port_pair_list: List[Tuple[str, int]],
+    listen_port: int,
+):
+    """
+    returns the node_instance of the dht
+    if no node instance has been created it
+    connects to the distributed hash table
+    if no bootstrap ip port pair list is given, it starts a new dht
+    :param bootstrap_ip_port_pair_list:
+    list of ip port tuples to connect to the dht
+    :param listen_port: port to listen on
+    :return: instance of server
+    """
+    global _node_instance
+    if _node_instance is None:
+        _node_instance = connect_dht(bootstrap_ip_port_pair_list, listen_port)
+
+    return _node_instance
+
+
+def connect_dht(
+    bootstrap_ip_port_pair_list: List[Tuple[str, int]],
+    listen_port: int,
+) -> Server:
+    """
+    connects to the distributed hash table
+    if no bootstrap ip port pair list is given, it starts a new dht
+    :param bootstrap_ip_port_pair_list:
+    list of ip port tuples to connect to the dht
+    :param listen_port: port to listen on
+    :return: instance of server
+    """
+    node = Server(storage=DropPeerDHTStorage(TRACKER_DROP_AVAILABILITY_TTL))
+    node.listen(listen_port)
+    loop = asyncio.get_event_loop()
+    if len(bootstrap_ip_port_pair_list) > 0:
+        loop.run_until_complete(node.bootstrap(bootstrap_ip_port_pair_list))
+
+    return node
+
+
+class DropPeerDHTStorage(IStorage):
+    def __init__(self, ttl: int=TRACKER_DROP_AVAILABILITY_TTL) -> None:
+        """
+        By default, max age is a week.
+        """
+        self.data = OrderedDict()  # type: OrderedDict
+        self.ttl = ttl
+
+    def __setitem__(self, key: bytes, value: Tuple[bytes, str, int]) -> None:
+        self.cull_entry(key)
+        if key in self.data:
+            self.data[key] = self.data[key] + [
+                (int(time.monotonic()), value),
+            ]
+        else:
+            self.data[key] = [
+                (int(time.monotonic()), value),
+            ]
+
+    def cull_entry(self, key: bytes) -> None:
+        if key not in self.data:
+            return
+        old_data = self.data[key]
+        new_data = []  # type: List[Tuple[bytes, str, int]]
+        for entry in old_data:
+            if (entry[0] + self.ttl > time.monotonic() and
+                    entry[0] < time.monotonic()):
+                new_data += entry
+
+        self.data[key] = new_data
+
+    def get(self, key: bytes, default=None) -> Any:
+        if key in self.data:
+            return self.__getitem__(key)
+        return default
+
+    def __getitem__(self, key: bytes) -> Any:
+        if key in self.data:
+            self.cull_entry(key)
+            return self[key]
+        raise Exception("Key not found")
```

Comment:
I think there's a specific exception that `__getitem__` should return? It looks like it should be `raise KeyError(key)`